File: c/test_yay.c
/*
 * YAY Parser Test Runner
 *
 * This program runs all test fixtures and compares parsed results
 * against expected values. It also tests error cases.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "yay.h"
#include "fixtures_gen.h"

/* ANSI color codes */
#define COLOR_RED     "\x1b[31m"
#define COLOR_GREEN   "\x1b[32m"
#define COLOR_YELLOW  "\x1b[33m"
#define COLOR_RESET   "\x1b[0m"

/* Test statistics */
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

/* Error test statistics */
static int error_tests_run = 0;
static int error_tests_passed = 0;
static int error_tests_failed = 0;

/* Print a value diff for debugging */
static void print_value_diff(const char *label, const yay_value_t *v) {
    char *str = yay_to_string(v);
    printf("  %s: %s\n", label, str ? str : "(null)");
    free(str);
}

/* Run a single valid test fixture */
static bool run_test(const test_fixture_t *fixture) {
    tests_run++;
    
    printf("Testing: %s ... ", fixture->name);
    fflush(stdout);
    
    /* Parse the YAY source */
    yay_result_t result = yay_parse(fixture->yay_source, 0, fixture->name);
    
    if (result.error) {
        printf(COLOR_RED "FAIL" COLOR_RESET " (parse error)\n");
        printf("  Error: %s\n", result.error->message);
        yay_result_free(&result);
        tests_failed++;
        return false;
    }
    
    /* Get expected value */
    yay_value_t *expected = fixture->make_expected();
    
    /* Compare */
    bool equal = yay_equal(result.value, expected);
    
    if (equal) {
        printf(COLOR_GREEN "PASS" COLOR_RESET "\n");
        tests_passed++;
    } else {
        printf(COLOR_RED "FAIL" COLOR_RESET " (value mismatch)\n");
        print_value_diff("Expected", expected);
        print_value_diff("Got     ", result.value);
        tests_failed++;
    }
    
    /* Cleanup */
    yay_free(expected);
    yay_result_free(&result);
    
    return equal;
}

/* Run a single error test fixture */
static bool run_error_test(const error_fixture_t *fixture) {
    error_tests_run++;
    
    printf("Testing: %s ... ", fixture->name);
    fflush(stdout);
    
    /* Parse the invalid YAY source - should fail */
    /* Use original_name to match expected error message format */
    yay_result_t result = yay_parse(fixture->nay_source, fixture->nay_len, fixture->original_name);
    
    if (!result.error) {
        printf(COLOR_RED "FAIL" COLOR_RESET " (expected error, got success)\n");
        char *str = yay_to_string(result.value);
        printf("  Got value: %s\n", str ? str : "(null)");
        free(str);
        yay_result_free(&result);
        error_tests_failed++;
        return false;
    }
    
    /* Check if error message contains expected pattern */
    if (strstr(result.error->message, fixture->error_pattern) != NULL) {
        printf(COLOR_GREEN "PASS" COLOR_RESET "\n");
        error_tests_passed++;
        yay_result_free(&result);
        return true;
    } else {
        printf(COLOR_RED "FAIL" COLOR_RESET " (error message mismatch)\n");
        printf("  Expected pattern: %s\n", fixture->error_pattern);
        printf("  Got: %s\n", result.error->message);
        yay_result_free(&result);
        error_tests_failed++;
        return false;
    }
}

/* Run all tests */
static void run_all_tests(void) {
    printf("\n");
    printf("========================================\n");
    printf("  YAY Parser C Test Suite\n");
    printf("========================================\n\n");
    
    printf("--- Valid Input Tests (.yay) ---\n\n");
    for (int i = 0; test_fixtures[i].name != NULL; i++) {
        run_test(&test_fixtures[i]);
    }
    
    printf("\n--- Error Tests (.nay) ---\n\n");
    for (int i = 0; error_fixtures[i].name != NULL; i++) {
        run_error_test(&error_fixtures[i]);
    }
    
    printf("\n");
    printf("========================================\n");
    printf("  Valid tests: %d/%d passed", tests_passed, tests_run);
    if (tests_failed > 0) {
        printf(" (" COLOR_RED "%d failed" COLOR_RESET ")", tests_failed);
    }
    printf("\n");
    printf("  Error tests: %d/%d passed", error_tests_passed, error_tests_run);
    if (error_tests_failed > 0) {
        printf(" (" COLOR_RED "%d failed" COLOR_RESET ")", error_tests_failed);
    }
    printf("\n");
    printf("  Total: %d/%d passed", tests_passed + error_tests_passed, tests_run + error_tests_run);
    if (tests_failed + error_tests_failed > 0) {
        printf(" (" COLOR_RED "%d failed" COLOR_RESET ")", tests_failed + error_tests_failed);
    }
    printf("\n");
    printf("========================================\n\n");
}

/* Run a specific test by name */
static bool run_named_test(const char *name) {
    for (int i = 0; test_fixtures[i].name != NULL; i++) {
        if (strcmp(test_fixtures[i].name, name) == 0) {
            return run_test(&test_fixtures[i]);
        }
    }
    
    printf("Unknown test: %s\n", name);
    return false;
}

/* List all available tests */
static void list_tests(void) {
    printf("Available tests:\n");
    for (int i = 0; test_fixtures[i].name != NULL; i++) {
        printf("  %s\n", test_fixtures[i].name);
    }
    printf("\nTotal: %d tests\n", TEST_FIXTURE_COUNT);
}

/* Parse and print a YAY file (for debugging) */
static void parse_file(const char *filename) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        fprintf(stderr, "Cannot open file: %s\n", filename);
        return;
    }
    
    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    char *content = malloc(size + 1);
    size_t nread = fread(content, 1, size, f);
    content[nread] = '\0';
    fclose(f);
    
    yay_result_t result = yay_parse(content, size, filename);
    
    if (result.error) {
        fprintf(stderr, "Parse error: %s\n", result.error->message);
    } else {
        char *str = yay_to_string(result.value);
        printf("%s\n", str);
        free(str);
    }
    
    yay_result_free(&result);
    free(content);
}

/* Print usage */
static void usage(const char *prog) {
    printf("Usage: %s [options]\n", prog);
    printf("\n");
    printf("Options:\n");
    printf("  (no args)       Run all tests\n");
    printf("  -l, --list      List all available tests\n");
    printf("  -t, --test NAME Run a specific test by name\n");
    printf("  -f, --file FILE Parse a YAY file and print result\n");
    printf("  -h, --help      Show this help\n");
}

int main(int argc, char *argv[]) {
    if (argc == 1) {
        /* No arguments - run all tests */
        run_all_tests();
        return (tests_failed + error_tests_failed) > 0 ? 1 : 0;
    }
    
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            usage(argv[0]);
            return 0;
        }
        else if (strcmp(argv[i], "-l") == 0 || strcmp(argv[i], "--list") == 0) {
            list_tests();
            return 0;
        }
        else if (strcmp(argv[i], "-t") == 0 || strcmp(argv[i], "--test") == 0) {
            if (i + 1 >= argc) {
                fprintf(stderr, "Missing test name\n");
                return 1;
            }
            bool passed = run_named_test(argv[++i]);
            return passed ? 0 : 1;
        }
        else if (strcmp(argv[i], "-f") == 0 || strcmp(argv[i], "--file") == 0) {
            if (i + 1 >= argc) {
                fprintf(stderr, "Missing filename\n");
                return 1;
            }
            parse_file(argv[++i]);
            return 0;
        }
        else {
            fprintf(stderr, "Unknown option: %s\n", argv[i]);
            usage(argv[0]);
            return 1;
        }
    }
    
    return 0;
}

File: python/test_yay.py
#!/usr/bin/env python3
"""
Test harness for the YAY parser.

Runs against all .yay fixture files and compares with expected .js output.
Runs against all .nay fixture files and verifies errors are raised.
"""

import json
import math
import os
import re
import sys

sys.path.insert(0, os.path.dirname(__file__))
import libyay as yay


def parse_js_value(js_content: str):
    """
    Parse a JavaScript value literal into Python equivalent.

    Handles:
    - BigInt literals (10n -> 10)
    - Infinity, -Infinity, NaN
    - Uint8Array.fromHex("...") -> bytes
    - new Uint8Array(0) -> b''
    - Arrays and objects
    """
    js_content = js_content.strip()

    # Handle special float values
    if js_content == "Infinity":
        return float("inf")
    if js_content == "-Infinity":
        return float("-inf")
    if js_content == "NaN":
        return float("nan")

    # Handle null
    if js_content == "null":
        return None

    # Handle booleans
    if js_content == "true":
        return True
    if js_content == "false":
        return False

    # Handle empty Uint8Array
    if js_content == "new Uint8Array(0)":
        return b""

    # Handle Uint8Array.fromHex("...")
    if js_content.startswith("Uint8Array.fromHex("):
        hex_str = js_content[20:-2]  # Extract hex string
        return bytes.fromhex(hex_str)

    # Handle BigInt literals (e.g., 10n, -10n)
    if (
        js_content.endswith("n")
        and not js_content.startswith('"')
        and not js_content.startswith("'")
    ):
        try:
            return int(js_content[:-1])
        except ValueError:
            pass

    # Handle negative numbers
    if js_content.startswith("-"):
        rest = js_content[1:]
        if rest.endswith("n"):
            return -int(rest[:-1])

    # Try parsing as JSON (handles strings, numbers, arrays, objects)
    try:
        # First, preprocess to handle BigInt in arrays/objects
        processed = preprocess_js_for_json(js_content)
        return json.loads(processed)
    except json.JSONDecodeError:
        pass

    # Handle plain integers
    try:
        return int(js_content)
    except ValueError:
        pass

    # Handle floats
    try:
        return float(js_content)
    except ValueError:
        pass

    raise ValueError(f"Cannot parse JS value: {js_content!r}")


def preprocess_js_for_json(js: str) -> str:
    """
    Preprocess JavaScript to be valid JSON.

    - Remove BigInt 'n' suffix
    - Quote unquoted object keys
    - Handle single-quoted strings
    """
    result = []
    i = 0
    in_string = False
    string_char = None

    while i < len(js):
        ch = js[i]

        if in_string:
            if ch == "\\" and i + 1 < len(js):
                result.append(ch)
                result.append(js[i + 1])
                i += 2
                continue
            if ch == string_char:
                in_string = False
                result.append('"')  # Always output double quote
                i += 1
                continue
            result.append(ch)
            i += 1
            continue

        if ch in "\"'":
            in_string = True
            string_char = ch
            result.append('"')  # Always output double quote
            i += 1
            continue

        # Handle BigInt suffix
        if ch == "n" and result and result[-1].isdigit():
            # Skip the 'n'
            i += 1
            continue

        result.append(ch)
        i += 1

    return "".join(result)


def values_equal(a, b) -> bool:
    """Compare two values, handling NaN specially."""
    if type(a) != type(b):
        # Special case: int vs float - compare as floats to handle large numbers
        # where int conversion loses precision (e.g., 6.022e23)
        if isinstance(a, int) and isinstance(b, float):
            return float(a) == b
        if isinstance(a, float) and isinstance(b, int):
            return a == float(b)
        return False

    if isinstance(a, float):
        if math.isnan(a) and math.isnan(b):
            return True
        return a == b

    if isinstance(a, dict):
        if set(a.keys()) != set(b.keys()):
            return False
        return all(values_equal(a[k], b[k]) for k in a)

    if isinstance(a, list):
        if len(a) != len(b):
            return False
        return all(values_equal(x, y) for x, y in zip(a, b))

    return a == b


def run_valid_tests(test_root: str, verbose: bool = False) -> tuple[int, int, list]:
    """
    Run all .yay tests from test_root/yay/, checking against test_root/js/.

    Returns (passed, failed, errors) where errors is a list of (filename, error_message).
    """
    passed = 0
    failed = 0
    errors = []

    yay_dir = os.path.join(test_root, "yay")
    js_dir = os.path.join(test_root, "js")
    yay_files = sorted(f for f in os.listdir(yay_dir) if f.endswith(".yay"))

    for fname in yay_files:
        yay_path = os.path.join(yay_dir, fname)
        js_path = os.path.join(js_dir, fname[:-4] + ".js")

        try:
            with open(yay_path, "rb") as fp:
                yay_bytes = fp.read()
            yay_content = yay_bytes.decode("utf-8")

            result = yay.loads(yay_content)

            # Check against expected output if .js file exists
            if os.path.exists(js_path):
                with open(js_path) as fp:
                    js_content = fp.read()

                try:
                    expected = parse_js_value(js_content)
                    if not values_equal(result, expected):
                        failed += 1
                        errors.append(
                            (
                                fname,
                                f"Value mismatch: got {result!r}, expected {expected!r}",
                            )
                        )
                        continue
                except ValueError as e:
                    # Can't parse expected value, just check parsing succeeded
                    if verbose:
                        print(f"  {fname}: Could not parse expected JS: {e}")

            passed += 1
            if verbose:
                print(f"  {fname}: OK -> {result!r}")

        except Exception as e:
            failed += 1
            errors.append((fname, str(e)))

    return passed, failed, errors


def run_error_tests(test_root: str, verbose: bool = False) -> tuple[int, int, list]:
    """
    Run all .nay error tests from test_root/nay/.

    Returns (passed, failed, errors) where errors is a list of (filename, error_message).
    """
    passed = 0
    failed = 0
    errors = []

    nay_dir = os.path.join(test_root, "nay")
    nay_files = sorted(f for f in os.listdir(nay_dir) if f.endswith(".nay"))

    for fname in nay_files:
        basename = fname[:-4]  # Remove .nay extension
        nay_path = os.path.join(nay_dir, fname)
        error_path = os.path.join(nay_dir, fname[:-4] + ".error")

        try:
            with open(nay_path, "rb") as fp:
                nay_bytes = fp.read()
            nay_content = nay_bytes.decode("utf-8")
        except UnicodeDecodeError:
            # Some .nay files may have invalid UTF-8 (like BOM test)
            # Read as latin-1 to get raw bytes as string
            with open(nay_path, "rb") as fp:
                nay_bytes = fp.read()
            nay_content = nay_bytes.decode("latin-1")

        # Read expected error substring
        expected_error = None
        if os.path.exists(error_path):
            with open(error_path) as fp:
                expected_error = fp.read().strip()

        try:
            result = yay.loads(nay_content)
            # Should have raised an error
            failed += 1
            errors.append((fname, f"Expected error but got result: {result!r}"))
        except yay.YayError as e:
            # Check if error message matches expected pattern
            error_msg = str(e)
            if expected_error:
                # Extract the key part of expected error (before "at X:Y of")
                # e.g., "Unexpected space after \":\"" from full message
                match = re.match(r"^(.+?)\s+at\s+\d+:\d+", expected_error)
                if match:
                    expected_pattern = match.group(1).strip()
                    if expected_pattern.lower() in error_msg.lower():
                        passed += 1
                        if verbose:
                            print(f"  {fname}: OK (error: {error_msg})")
                    else:
                        failed += 1
                        errors.append(
                            (
                                fname,
                                f"Error mismatch: got '{error_msg}', expected pattern '{expected_pattern}'",
                            )
                        )
                else:
                    # No "at X:Y" pattern, just check if any error was raised
                    passed += 1
                    if verbose:
                        print(f"  {fname}: OK (error: {error_msg})")
            else:
                passed += 1
                if verbose:
                    print(f"  {fname}: OK (error: {error_msg})")
        except Exception as e:
            # Other exceptions count as pass (we expected an error)
            passed += 1
            if verbose:
                print(f"  {fname}: OK (exception: {type(e).__name__}: {e})")

    return passed, failed, errors


def main():
    import argparse

    parser = argparse.ArgumentParser(description="Test YAY parser against fixtures")
    parser.add_argument(
        "-v", "--verbose", action="store_true", help="Show all test results"
    )
    parser.add_argument("-f", "--file", help="Test a specific .yay file")
    parser.add_argument(
        "--test-dir", default="../test", help="Root directory containing test fixtures"
    )
    args = parser.parse_args()

    if args.file:
        # Test a single file
        try:
            with open(args.file, "rb") as fp:
                content_bytes = fp.read()
            try:
                content = content_bytes.decode("utf-8")
            except UnicodeDecodeError:
                content = content_bytes.decode("latin-1")
            print(f"Input:\n{content}")
            print(f"\nParsing...")
            result = yay.loads(content)
            print(f"Result: {result!r}")
            print(f"Type: {type(result).__name__}")
        except Exception as e:
            print(f"ERROR: {e}")
            import traceback

            traceback.print_exc()
            sys.exit(1)
    else:
        # Run all tests
        test_root = os.path.join(os.path.dirname(__file__), args.test_dir)
        print(f"Running tests from {test_root}\n")

        # Run valid .yay tests
        print("Valid input tests (.yay):")
        v_passed, v_failed, v_errors = run_valid_tests(test_root, args.verbose)
        print(f"  {v_passed} passed, {v_failed} failed")

        # Run error .nay tests
        print("\nError tests (.nay):")
        e_passed, e_failed, e_errors = run_error_tests(test_root, args.verbose)
        print(f"  {e_passed} passed, {e_failed} failed")

        total_passed = v_passed + e_passed
        total_failed = v_failed + e_failed
        all_errors = v_errors + e_errors

        print(f"\nTotal: {total_passed} passed, {total_failed} failed")

        if all_errors:
            print("\nFailures:")
            for fname, err in all_errors:
                print(f"  {fname}: {err}")

        sys.exit(0 if total_failed == 0 else 1)


if __name__ == "__main__":
    main()
